clang/tools/clang-scan-deps/clang-deps-launcher.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+"""
+Launcher for clang which computes the dependency for the file being compiled.
+"""
+import sys
+import os
+import subprocess
+import time
+import json
+import tempfile
+
+
+TOOLS_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+def trim_clang_args_in_cdb(cdb):
+    """Trim -Xclang options away since they are not portable"""
+    # We also replace the clang exe path so it can find the clang headers.
+    clang_exe = os.path.basename(cdb['arguments'][0])
+    trim_args = [os.path.join(TOOLS_DIR, clang_exe)]
+    skip = False
+    for arg in cdb['arguments'][1:]:
+        if skip:
+            skip = False
+            continue
+        if arg == '-Xclang':
+            skip = True
+            continue
+        trim_args.append(arg)
+    cdb['arguments'] = trim_args
+
+
+def main():
+    # Clang command is the input.
+    clang_cmd = sys.argv[1:]
+    # Figure out the output path.
+    output_idx = len(clang_cmd) - clang_cmd[::-1].index('-o')
+    output_object = clang_cmd[output_idx]
+    # Write compilation data base to STDOUT, and strip the last ',' so it is a
+    # valid JSON object.
+    clang_cmd.extend(['-MJ', '-'])
+    start = time.perf_counter()
+    cdb = subprocess.check_output(clang_cmd).strip()[:-1]
+    end = time.perf_counter()
+    output_size = os.path.getsize(output_object)
+
+    # load and clean up compiler commands, write to a temp file.
+    cdb_object = json.loads(cdb)
+    trim_clang_args_in_cdb(cdb_object)
+
+    cdb_output = [ cdb_object ]
+    cdb_file = tempfile.NamedTemporaryFile(suffix='.cdb', mode='w+')
+    json.dump(cdb_output, cdb_file)
+    cdb_file.flush()
+
+    # run clang-scan-deps to collect dependency.
+    scan_deps = os.path.join(TOOLS_DIR, "clang-scan-deps")
+    scan_deps_cmd = [scan_deps, '-compilation-database', cdb_file.name,
+                     '-format', 'experimental-tree-full']
+
+    scan_deps_output = subprocess.check_output(scan_deps_cmd)
+    cdb_file.close()
```

Comment:
Is there  a reason to not close this immediately after writing it?

Reply:
I didn't pass the flag to Tempfile that will keep the file around so closing will delete the file. Thus `flush` is needed before clang-scan-deps can be executed.